File: baekjoon/java/14923.java
import java.io.*;
import java.util.*;

public class Main {

    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

    static int Row, Col, sXpos, sYpos, eXpos, eYpos;
    static int[][] matrix;

    static boolean[][][] visited;
    static int[][][] dist;

    static int[] xMove = {-1, 1, 0, 0};
    static int[] yMove = {0, 0, -1, 1};

    public static void main(String[] args) throws Exception {
        input();
        int ans = bfs();
        System.out.println(ans);
    }

    static boolean inTheMatrix(int xpos, int ypos) {
        return xpos >= 1 && xpos <= Row && ypos >= 1 && ypos <= Col;
    }

    static int bfs() {
        ArrayDeque<int[]> q = new ArrayDeque<>();
        q.offer(new int[]{0, sXpos, sYpos});
        visited[0][sXpos][sYpos] = true;
        dist[0][sXpos][sYpos] = 0;

        while (!q.isEmpty()) {
            int[] cur = q.poll();
            int b = cur[0];
            int r = cur[1];
            int c = cur[2];

            if (r == eXpos && c == eYpos) {
                return dist[b][r][c];
            }

            for (int i = 0; i < 4; i++) {
                int nr = r + xMove[i];
                int nc = c + yMove[i];
                if (!inTheMatrix(nr, nc)) continue;

                if (matrix[nr][nc] == 0)
                {
                    if (!visited[b][nr][nc]) {
                        visited[b][nr][nc] = true;
                        dist[b][nr][nc] = dist[b][r][c] + 1;
                        q.offer(new int[]{b, nr, nc});
                    }
                }
                else
                {
                    if (b == 0 && !visited[1][nr][nc]) {
                        visited[1][nr][nc] = true;
                        dist[1][nr][nc] = dist[0][r][c] + 1;
                        q.offer(new int[]{1, nr, nc});
                    }
                }
            }
        }

        return -1;
    }

    static void input() throws Exception {
        StringTokenizer st = new StringTokenizer(br.readLine());
        Row = Integer.parseInt(st.nextToken());
        Col = Integer.parseInt(st.nextToken());

        st = new StringTokenizer(br.readLine());
        sXpos = Integer.parseInt(st.nextToken());
        sYpos = Integer.parseInt(st.nextToken());

        st = new StringTokenizer(br.readLine());
        eXpos = Integer.parseInt(st.nextToken());
        eYpos = Integer.parseInt(st.nextToken());

        matrix = new int[Row + 1][Col + 1];

        visited = new boolean[2][Row + 1][Col + 1];
        dist = new int[2][Row + 1][Col + 1];

        for (int i = 1; i <= Row; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 1; j <= Col; j++) {
                matrix[i][j] = Integer.parseInt(st.nextToken());
            }
        }
    }
}

File: baekjoon/C++/14923.cc
#include <iostream>
#include <vector>
#include <algorithm>
#include <queue>
#define endl '\n'
using namespace std;
using pos = pair<int,int>;

int row, col;
pos start, target;
vector<vector<int>> matrix;
vector<vector<vector<int>>> dis;
vector<vector<vector<bool>>> visited;

const int x_move[] = {-1,1,0,0}, y_move[] = {0,0,-1,1};

template<typename matrix> 
void show_matrix(const matrix& m) 
{
    for (int i = 1; i <= row; i++)
    {
        for (int j = 1; j <= col; j++)
        {
            cout << m[i][j] << " ";
        } cout << endl;
    }
}

void input() noexcept 
{
    cin >> row >> col;
    cin >> start.first >> start.second;
    cin >> target.first >> target.second;
    
    // init container
    matrix.assign(row+1, vector<int>(col+1));
    dis.assign(2, vector<vector<int>>(row+1, vector<int>(col+1)));
    visited.assign(2, vector<vector<bool>>(row+1, vector<bool>(col+1)));
    
    for (int i = 1; i <= row; i++)
    {
        for (int j = 1; j <= col; j++)
        {
            cin >> matrix[i][j];
        }
    }
}

bool inTheMatrix(const int& xpos, const int& ypos) {
    return xpos >= 1 && xpos <= row && ypos >= 1 && ypos <= col;
}

void bfs()
{
    queue<pair<int, pos>> q; // {h, (x,y)}
    q.push({0, start});
    visited[0][start.first][start.second] = true;
    dis[0][start.first][start.second] = 0;

    while (!q.empty()) {
        auto [h, p] = q.front();
        q.pop();

        int x = p.first, y = p.second;

        if (x == target.first && y == target.second) {
            cout << dis[h][x][y] << endl;
            return;
        }

        for (int i = 0; i < 4; ++i) {
            int nx = x + x_move[i];
            int ny = y + y_move[i];
            if (!inTheMatrix(nx, ny)) continue;

            if (matrix[nx][ny] == 0) {
                if (visited[h][nx][ny]) continue;
                visited[h][nx][ny] = true;
                dis[h][nx][ny] = dis[h][x][y] + 1;
                q.push({h, {nx, ny}});
            } else { 
                // 벽이면: 아직 안 부쉈을 때만 부수고 이동
                if (h == 1) continue;
                if (visited[1][nx][ny]) continue;
                visited[1][nx][ny] = true;
                dis[1][nx][ny] = dis[0][x][y] + 1;
                q.push({1, {nx, ny}});
            }
        }
    }

    cout << -1 << endl;
}


int main(void) {
    ios::sync_with_stdio(0);
    cin.tie(0); cout.tie(0);


    input();
    bfs();

    return (0);
}